### DISCUSSION TITLE
Sort uppercase completions to the bottom

### DISCUSSION THREAD

**Discussion by Crandel** [company-mode/company-mode#1190] [Q&A] [answered]
my company-backends is 
`
company-backends '((company-yasnippet company-files company-semantic company-css 
                                    company-capf company-etags company-keywords company-nxml 
                                    company-dabbrev-code company-dabbrev))
`

When I press characters I get list of the found words, but it always started with capital letters
![screensho-21032016](https://cloud.githubusercontent.com/assets/4005032/15218521/022e008e-1869-11e6-9a17-0c60fe7e5f52.png)

Can you add some option to set capitalized words to the end of the list, because I use them very rarely.
I know that there are option
`company-dabbrev-downcase`
but I have many camelCase words in my code, so I set it to nil

**Comment by dgutov** [comment]
Hi!

How about this? You can write your own sorting predicate.

``` diff
diff --git a/company.el b/company.el
index 23ed09a..0af8236 100644
--- a/company.el
+++ b/company.el
@@ -454,6 +454,9 @@ even if the backend uses the asynchronous calling convention."

 (put 'company-backends 'safe-local-variable 'company-safe-backends-p)

+(defcustom company-sort-predicate #'string<
+  "Function to sort the candidates if the backend returns them unsorted.")
+
 (defcustom company-transformers nil
   "Functions to change the list of candidates received from backends.

@@ -1203,7 +1206,7 @@ can retrieve meta-data for them."

 (defun company--preprocess-candidates (candidates)
   (unless (company-call-backend 'sorted)
-    (setq candidates (sort candidates 'string<)))
+    (setq candidates (sort candidates company-sort-predicate)))
   (when (company-call-backend 'duplicates)
     (setq candidates (company--strip-duplicates candidates)))
   candidates)

```

**Comment by Crandel** [comment]
Thank you for help. Can you create a small simple example how it can be at .emacs file, I couldn`t imagine how this can be implemented

**Comment by dgutov** [comment, marked as answer]
OK, actually forget the patch. Try this:

``` el
(defun my-sort-uppercase (candidates)
  (let (case-fold-search
        (re "\\`[[:upper:]]*\\'"))
    (sort candidates
          (lambda (s1 s2)
            (and (string-match-p re s2)
                 (not (string-match-p re s1)))))))

(push 'my-sort-uppercase company-transformers)
```

Not dabbrev-specific, it will sort all completions with uppercase symbols at the end.

**Comment by Crandel** [comment]
Thank you so much, you really made my work easy
